File: src/com/dynatrace/webautomation/DynaTraceWebDriver.java
package com.dynatrace.webautomation;

import java.net.URISyntaxException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class DynaTraceWebDriver implements WebDriver {

	private WebDriver driver;
	private DynaTraceWebDriverNavigation dynaTraceNavigation;
	private static HashMap<WebDriver, DynaTraceWebDriver> drivers = new HashMap<WebDriver, DynaTraceWebDriver>();
	
	public DynaTraceWebDriver(WebDriver driver) {
		this.driver = driver;
		this.dynaTraceNavigation = new DynaTraceWebDriverNavigation(driver);
	}
	
	public static DynaTraceWebDriver forWebDriver(WebDriver driver) {
		DynaTraceWebDriver dtDriver = drivers.get(driver);
		if(dtDriver == null) {
			dtDriver = new DynaTraceWebDriver(driver);
			drivers.put(driver, dtDriver);
		}
		return dtDriver;
	}
	
	@Override
	public void close() {
		driver.close();
	}

	@Override
	public WebElement findElement(By arg0) {
		return driver.findElement(arg0);
	}

	@Override
	public List<WebElement> findElements(By arg0) {
		return driver.findElements(arg0);
	}

	@Override
	public void get(String arg0) {
		driver.get(arg0);
	}

	@Override
	public String getCurrentUrl() {
		return driver.getCurrentUrl();
	}

	@Override
	public String getPageSource() {
		return driver.getPageSource();
	}

	@Override
	public String getTitle() {
		return driver.getTitle();
	}

	@Override
	public String getWindowHandle() {
		return driver.getWindowHandle();
	}

	@Override
	public Set<String> getWindowHandles() {
		return driver.getWindowHandles();
	}

	@Override
	public Options manage() {
		return driver.manage();
	}

	@Override
	public Navigation navigate() {
		return dynaTraceNavigation;
	}

	@Override
	public void quit() {
		driver.quit();
	}

	@Override
	public TargetLocator switchTo() {
		return driver.switchTo();
	}

	private class DynaTraceWebDriverNavigation implements WebDriver.Navigation
	{
		protected WebDriver driver;
		protected DynaTraceWebDriverHelper dynaTrace;
		public DynaTraceWebDriverNavigation(WebDriver driver) {
			this.driver = driver;
			dynaTrace = DynaTraceWebDriverHelper.forDriver(driver);
		}
		
		@Override
		public void back() {
			driver.navigate().back();
		}

		@Override
		public void forward() {
			driver.navigate().forward();
			
		}

		@Override
		public void refresh() {
			driver.navigate().refresh();
		}

		@Override
		public void to(String arg0) {
			if(!dynaTrace.hasTimerName())
				dynaTrace.setTimerName(true);
			driver.navigate().to(dynaTrace.getDynaTraceTimerUrl(arg0));
		}

		@Override
		public void to(URL arg0) {
			try {
				to(arg0.toURI().toString());
			} catch (URISyntaxException e) {
			}
		}
		
	}

	public WebDriver getWebDriver() {
		return driver;
	}
}

File: dotnet/src/DynaTraceWebDriver.cs
using OpenQA.Selenium;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace DynatraceCSharpHelper
{
    public class DynaTraceWebDriver : IWebDriver
    {
        private IWebDriver driver;
        private DynaTraceWebDriverNavigation dynaTraceNavigation;
        private static Dictionary<IWebDriver, DynaTraceWebDriver> drivers = new Dictionary<IWebDriver, DynaTraceWebDriver>();

        public DynaTraceWebDriver(IWebDriver driver)
        {
            this.driver = driver;
            this.dynaTraceNavigation = new DynaTraceWebDriverNavigation(driver);
        }

        public static DynaTraceWebDriver forWebDriver(IWebDriver driver)
        {
            DynaTraceWebDriver dtDriver = null;

            drivers.TryGetValue(driver, out dtDriver);
            if (dtDriver == null)
            {
                dtDriver = new DynaTraceWebDriver(driver);
                drivers.Add(driver, dtDriver);
            }
            return dtDriver;
        }

        public string CurrentWindowHandle
        {
            get
            {
                return driver.CurrentWindowHandle;
            }
        }

        public string PageSource
        {
            get
            {
                return driver.PageSource;
            }
        }

        public string Title
        {
            get
            {
                return driver.Title;
            }
        }

        public string Url
        {
            get
            {
                return driver.Url;
            }

            set
            {
                driver.Url = Url;
            }
        }

        public ReadOnlyCollection<string> WindowHandles
        {
            get
            {
                return driver.WindowHandles;
            }
        }

        public void Close()
        {
            driver.Close();
        }

        public void Dispose()
        {
            driver.Dispose();
        }

        public IWebElement FindElement(By by)
        {
            return driver.FindElement(by);
        }

        public ReadOnlyCollection<IWebElement> FindElements(By by)
        {
            return driver.FindElements(by);
        }

        public IOptions Manage()
        {
            return driver.Manage();
        }

        public INavigation Navigate()
        {
            return driver.Navigate();
        }

        public void Quit()
        {
            driver.Quit();
        }

        public ITargetLocator SwitchTo()
        {
            return driver.SwitchTo();
        }

        private class DynaTraceWebDriverNavigation : INavigation
        {
            protected IWebDriver driver;
            protected DynaTraceWebDriverHelper dynaTrace;
            public DynaTraceWebDriverNavigation(IWebDriver driver)
            {
                this.driver = driver;
                dynaTrace = DynaTraceWebDriverHelper.forDriver(driver);
            }


            public void Back()
            {
                driver.Navigate().Back();
            }

            public void Forward()
            {
                driver.Navigate().Forward();
            }

            public void GoToUrl(Uri url)
            {                              
                    driver.Navigate().GoToUrl(url);             
            }

            public void GoToUrl(string url)
            {
                try
                {
                    driver.Navigate().GoToUrl(url);
                }
                catch (UriFormatException)
                {   
                }
            }

            public void Refresh()
            {
                driver.Navigate().Refresh();
            }
        }

        public IWebDriver getWebDriver()
        {
            return driver;
        }
    }

   
}